File: daily/cpp/1750_min_len_string_after_deleting_similar_ends.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
 public:
  /* Two pointer w/ one traversal */
  int minimumLength(string s) {
    int le = 0, ri = s.size() - 1;
    while (le < ri && s[le] == s[ri]) {
      char c = s[le];
      while (le <= ri && s[le] == c) ++le;
      while (le <= ri && s[ri] == c) --ri;
    }
    return ri - le + 1;
  }
};

File: daily/python/1750_min_len_string_after_deleting_similar_ends.py
class Solution:
    def minimumLength(self, s: str) -> int:
        # this is not a good solution
        # while len(s) > 1 and s[0] == s[-1]:
        #     c = s[0]
        #     while s and s[0] == c:
        #         s = s[1:]
        #     while s and s[-1] == c:
        #         s = s[:-1]
        # return len(s)

        # better solution
        le = 0
        ri = len(s) - 1
        while le < ri and s[le] == s[ri]:
            c = s[le]
            while le <= ri and s[le] == c:
                le += 1
            while le <= ri and s[ri] == c:
                ri -= 1
        return ri - le + 1